### ISSUE TITLE
Fails to empty the directory on Macos if the directory contains subdirectories.

### ISSUE THREAD

**Issue by hkratz** [XAMPPRocky/remove_dir_all#35]
`ensure_empty_dir()` and `remove_dir_contents()` fail on Macos whenever the passed directory contains subdirectories because on Macos unlink(2) returns EPERM instead of EISDIR when called on a directory.

From the `unlink(2)` man page on Macos :
```
    [EPERM]            The named file is a directory and the effective user ID of the
                        process is not the super-user.
```

**Comment by rmustacc** [comment]
Hi there, I got here in a bit of a roundabout way via rustup. While writing up the details, I then found this issue. So I just want to add to what @hkratz already wrote up and how this could be hit by folks via rustup on another platform and the debugging steps I took in case someone else ends up here.

The short form is that it appears the conditions that are used to determine if we should proceed into calling `fs::remove_dir_all` in `remove_file_or_dir_all` may not be quite all of them that different systems can hit. Let me provide a bit more background of how we ended up here.

I'm using `rustup` and tried to do a `rustup` update that failed:

```
rm@atrium ~ $ RUST_BACKTRACE=1 rustup update
info: syncing channel updates for 'stable-x86_64-unknown-illumos'
info: checking for self-updates

  stable-x86_64-unknown-illumos unchanged - rustc 1.65.0 (897e37553 2022-11-02)

info: cleaning up downloads & tmp directories
thread 'main' panicked at 'Unable to clean up /home/rm/.rustup/tmp: Os { code: 1, kind: PermissionDenied, message: "Not owner" }', src/utils/utils.rs:649:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: rustup::utils::utils::delete_dir_contents
   3: rustup::cli::rustup_mode::update
   4: rustup::cli::rustup_mode::main
   5: rustup_init::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I wanted to confirm the file that we were removing and with truss (similar to strace) saw:

```
...
/1:     unlink("/home/rm/.rustup/tmp/euewxq3ynqp1ya9c_dir") Err#1 EPERM
...
```

And if we look at this in the file system itself, we see it was a directory. When I first saw an `EPERM`, I wondered if I somehow didn't own the directory in question; however, that's not the case:

```
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
rm@atrium ~ $ ls -l ~/.rustup/tmp/
total 4
drwxr-xr-x   3 rm       staff          7 Oct 12 16:05 euewxq3ynqp1ya9c_dir
-rw-r--r--   1 rm       staff         29 Oct 12 16:05 rcj_uw1k61zz75q3_file
```

In this case, the `EPERM` error appear to be due to the fact that we're trying to remove a directory via `unlink`. This is due to the fact that historically unlinking a directory was supported for privileged users on the UFS file system though it was always a bad idea. I'm not a privileged user, which is why we're getting this and in general this is being removed, but the error code at least for today's systems is still not going to match the current check for `libc::EISDIR`.

**Comment by XAMPPRocky** [comment]
@rmustacc Thank you for investigating the issue! This will definitely be helpful in writing the test case which was mentioned as one of the things required for merging #36

**Comment by rmustacc** [comment]
@XAMPPRocky Tese case written for you in #49.

**Comment by XAMPPRocky** [comment]
@rmustacc Thank you for your PR! Do you know happen to know what the more modern syscall that we should be using for these systems would be?

**Comment by rmustacc** [comment]
There's not really another system call to use here. Basically if you have a directory you want to use what you're already using, that is `fs::remove_dir_all`. However, that rust function doesn't work on files. There is on system call that basically implements rm -rf directory in most *nix systems. So either you want to use something like the pattern that @hkratz already did in #36 and check on failure, or use openat(2), fstat(2), and walk and remove things directly, but probably better to stick with `fs::remove_dir_all` than trying to reinvent that. There's a TOCTTOU case with this, but I think that's somewhat inevitable.

**Comment by rbtcollins** [comment]
This is now fixed in 0.8.1 and above
